contracts/core/MarginPoolV2.sol

```diff
@@ -0,0 +1,244 @@
+    /// @notice emit event when a borrower borrows an asset
+    event Borrow(address indexed oToken, address indexed collateralAsset, uint256 amount, address indexed borrower);
+    /// @notice emit event when a loan is repaid
+    event Repay(
+        address indexed oToken,
+        address indexed collateralAsset,
+        uint256 amount,
+        address indexed borrower,
+        address repayer
+    );
+
+    /**
+     * @notice check if the sender is whitelisted
+     */
+    modifier onlyWhitelistedBorrower() {
+        require(whitelistedBorrower[msg.sender], "MarginPool: Sender is not whitelisted borrower");
+
+        _;
+    }
+
+    /**
+     * @notice Set borrower whitelist status
+     * @param _borrower address of the borrower (100% of the time verified market makers)
+     * @param _whitelisted bool of whether it is whitelisted or blacklisted
+     */
+    function setBorrowerWhitelistedStatus(address _borrower, bool _whitelisted) external onlyOwner {
+        require(_borrower != address(0), "!_borrower");
+
+        whitelistedBorrower[_borrower] = _whitelisted;
+        emit SetBorrowWhitelist(_borrower, _whitelisted);
+    }
+
+    /**
+     * @notice Set ribbon vault whitelist status
+     * @param _ribbonVault address of the ribbon vault
+     * @param _whitelisted bool of whether it is whitelisted or blacklisted
+     */
+    function setRibbonVaultWhitelistedStatus(address _ribbonVault, bool _whitelisted) external onlyOwner {
+        require(_ribbonVault != address(0), "!_ribbonVault");
+
+        whitelistedRibbonVault[_ribbonVault] = _whitelisted;
+        emit SetRibbonVaultWhitelist(_ribbonVault, _whitelisted);
+    }
+
+    /**
+     * @notice Set buyer whitelist status
+     * @param _buyer address of the oToken buyer
+     * @param _whitelisted bool of whether it is whitelisted or blacklisted
+     */
+    function setBuyerWhitelistedStatus(address _buyer, bool _whitelisted) external onlyOwner {
+        require(_buyer != address(0), "!_buyer");
+
+        whitelistedOTokenBuyer[_buyer] = _whitelisted;
+        emit SetOTokenBuyerWhitelist(_buyer, _whitelisted);
+    }
+
+    /**
+     * @notice Set Borrow percent of collateral asset
+     * @param _collateral address of collateral asset
+     * @param _borrowPCT borrow PCT
+     */
+    function setBorrowPCT(address _collateral, uint256 _borrowPCT) external onlyOwner {
+        borrowPCT[_collateral] = _borrowPCT;
+        emit SetBorrowPCT(_collateral, _borrowPCT);
+    }
+
+    /**
+     * @notice Lends out asset to market maker
+     * @param _oToken address of the oToken laying claim to the collateral assets
+     * @param _amount amount of the asset to lend to borrower
+     */
+    function borrow(address _oToken, uint256 _amount) public onlyWhitelistedBorrower {
+        require(
+            WhitelistInterface(AddressBookInterface(addressBook).getWhitelist()).isWhitelistedOtoken(_oToken),
```

Comment:
Check the Whitelist if the otoken is created. This means they have to create the otokens via the factory and prevents fake otokens to drain some underlying